frame/balances/src/lib.rs

```diff
@@ -369,6 +369,33 @@ pub mod pallet {
 			<Self as Currency<_>>::transfer(&transactor, &dest, value, KeepAlive)?;
 			Ok(().into())
 		}
+
+		/// Transfer the entire transferable balance from the caller account.
+		///
+		/// The dispatch origin of this call must be Signed.
```

Comment:
Maybe not appropriate for here, but should there be some guidance to users of extrinsics like this on how provider/consumer ref counters may affect an attempt to reap an account?

I personally don't fully understand the use case of them.

In this case it looks like in `Balances::reducible_balance` we check
```
if frame_system::Pallet::<T>::can_dec_provider(who) && !keep_alive 
```
So as a user I would like to know in what situations  I _can't_ decrement the provider ref

Reply:
this function will not always reap an account. It will simply attempt to transfer as much transferable balance as possible.

Indeed other systems like democracy or staking (locks, reserves, ref counters) can affect a users access to transfer their full balance, but there is nothing that can be done about this at the pallet level.

Think about this tool as a perfect thing for exchanges to use when you send a balance to their "temp wallets".

I send some amount from A -> B, and want to send the full B -> C right after.

B will have nothing on the account except this balance which it wants to transfer, and this function is perfect for that.

Reply:
My concern is around confusion for when a user tries to use this to reap an account, but they only send `free - ED` bc they have a `consumer > 0`. (I suspect having to deal with refs on an account is a new paradigm for the average user?).

So I guess my concrete suggestion here would be to elaborate on what qualifies as `the entire transferable balance`. i.e

```
// Transfer the entire transferable balance from the caller account. 
// NOTE: Valid transferrable balance does not include `AcountInfo.reserved` funds. Further, if the 
// caller has `AcountInfo.consumer > 1`, only transfer `AcountInfo.free - exstential_deposit` will be transferred.
```

Reply:
I will add a bit more docs here, but by definition, "transferable balance" is defined as the balance which is allowed to be transferred.

This means no reserved balance, no locked balance, and no ED if outstanding references exist.

But yes, clarification of these points is worth having.